JSDK-2531 reduce build times by running tests on multiple containers.

makarandp0 opened the pull request:
This PR uses circleCI feature to enable [splitting tests](https://circleci.com/docs/2.0/parallelism-faster-jobs/) into multiple containers using timing information. [Example Run](https://circleci.com/workflow-run/9ea62eda-31a8-4415-b2c2-5efeaa3c9158) shows that runtime is reduced about 30%, and not its mostly by slowest running test files that is `connect.js` - Breaking up this file into multiple units would reduce overall CI time substantially.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

manjeshbhargav commented on scripts/karma.js:
```diff
@@ -29,32 +24,13 @@ function getTestPaths(path) {
-      console.warn(`You are splitting ${paths.length} files into ${totalRuns} groups!`);
-    }
+  if (process.env.TEST_FILES) {
+    console.log('test files provided were: ', process.env.TEST_FILES);
```
@makarandp0 Remove console log before merging.

manjeshbhargav commented on scripts/karma.js:
```diff
@@ -29,32 +24,13 @@ function getTestPaths(path) {
+    console.log('test files provided were: ', process.env.TEST_FILES);
+    let testFiles = process.env.TEST_FILES.split('\n');
+    testFiles = testFiles.map(file => resolvePath(file));
+    return paths.filter((path) => testFiles.includes(path));
```
```suggestion
    return paths.filter(path => testFiles.includes(path));
```

manjeshbhargav commented on scripts/circleci-run-tests.sh:
```diff
@@ -18,6 +18,10 @@ then
 else
+    # ask circleci to split tests by timing.
+    echo "Asking circleci to pick tests based on timings..."
+    export TEST_FILES=$(circleci tests glob "$PWD/test/integration/spec/**/*.js" | circleci tests split --split-by=timings)
```
@makarandp0 Does this make sure that test files are split evenly based on execution time?

makarandp0 replied:
Yes, that what the documentation seem to imply 

> On each successful run of a test suite, CircleCI saves timings data to a directory specified by the path in the store_test_results step. If you do not use store_test_results, there will be no timing data available for splitting your tests.
> To split by test timings, use the --split-by flag with the timings split type.

So far I have not seen it doing good Job. I was hoping to see connect.js not clubbed with other files, but I suspect that's because it does not have sufficient timing data to go by. But even if the split was not timing based, it seems to do a reasonable job.

manjeshbhargav commented:
@makarandp0 As part of this PR, make sure that the link to the Circle CI PR workflow appears just below the `continuous-integration/travis-ci/pr` link in the "All checks have passed" section. Also, please convert the README.md badge from travis to circle.

manjeshbhargav commented:
@makarandp0 Also it looks like there are too many commits for just 3 files changed. Please make sure your branch is synced with the changes from master.

makarandp0 commented:
@manjeshbhargav, I have now updated readme to show circleci badge, and also updated the settings to show circleCI results in pull request checks. Will merge this now.
